### PR TITLE
[red-knot] Deeper understanding of `LiteralString`

### LINKED ISSUE #14648: [red-knot] Narrowing on `s == "foo"` where `s` is of type `LiteralString`
For any symbol `s` of type `LiteralString`, it should be safe to narrow it according to comparisons of the pattern `s == "..."`.

```python
a: LiteralString

if a == "foo":
	reveal_type(a)  # revealed: Literal["foo"]
```

This issue is part of #13694.

### PR DESCRIPTION
**Pull request by InSyncWithFoo**
## Summary

Resolves #14648.

## Test Plan

Markdown tests.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types.rs [hunk 1]
```diff
@@ -629,6 +629,11 @@ impl<'db> Type<'db> {
                 Type::StringLiteral(_) | Type::LiteralString,
                 Type::Instance(InstanceType { class }),
             ) if class.is_known(db, KnownClass::Str) => true,
+            (
+                Type::Instance(InstanceType { class }),
+                Type::StringLiteral(_) | Type::LiteralString,
+            ) if class.is_known(db, KnownClass::Str) => false,
+            (Type::LiteralString, Type::StringLiteral(_)) => false,
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:636]: All of the tests in this PR pass without adding these lines. The default fallback case for `is_subtype_of` is "not a subtype", so generally we don't need to add specific `false` cases like this, unless they need to be handled before a subsequent "true" case.
```suggestion
```

File: crates/red_knot_python_semantic/src/types.rs [hunk 2]
```diff
@@ -1403,7 +1408,7 @@ impl<'db> Type<'db> {
             // `Any` is callable, and its return type is also `Any`.
             Type::Any => CallOutcome::callable(Type::Any),
 
-            Type::Todo(_) => CallOutcome::callable(todo_type!()),
+            Type::Todo(_) => CallOutcome::callable(todo_type!("call type")),
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types.rs:1411]: "call type" doesn't really make sense here, "call todo" would make more sense
```suggestion
            Type::Todo(_) => CallOutcome::callable(todo_type!("call todo")),
```

File: crates/red_knot_python_semantic/src/types/narrow.rs [hunk 3]
```diff
@@ -291,8 +291,15 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
...
         let mut constraints = NarrowingConstraints::default();
+
+        let mut last_rhs_ty: Option<Type> = None;
+
         for (op, (left, right)) in std::iter::zip(&**ops, comparator_tuples) {
+            let lhs_ty = last_rhs_ty.unwrap_or_else(|| {
+                inference.expression_ty(left.scoped_expression_id(self.db, scope))
+            });
             let rhs_ty = inference.expression_ty(right.scoped_expression_id(self.db, scope));
+            last_rhs_ty = Some(rhs_ty);
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/narrow.rs:302]: This looks like it's improving support for chained comparisons? It doesn't appear directly connected to `LiteralString`.

I'm not immediately seeing what test added in this PR requires this support?

Unless this is definitely required in this PR for reasons I'm not seeing, I think it would be easier to evaluate in a separate PR, accompanied by tests which specifically require this?

File: crates/red_knot_python_semantic/src/types/narrow.rs [hunk 4]
```diff
@@ -327,6 +334,12 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
                                 constraints.insert(symbol, ty);
                             }
                         }
+                        ast::CmpOp::Eq if lhs_ty.is_literal_string() => {
+                            let ty = IntersectionBuilder::new(self.db)
+                                .add_positive(rhs_ty)
+                                .build();
+                            constraints.insert(symbol, ty);
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/narrow.rs:341]: Building an intersection with just one type in it is a no-op, just use the single type directly.

```suggestion
                            constraints.insert(symbol, rhs_ty);
```

**Review by carljm — CHANGES_REQUESTED** [review]
Thanks for the contribution! I appreciate the thorough tests, but it is possible to have too many tests :) We should make sure that each test, and each part of each test, is asserting something meaningful about the added feature. Usually added tests should fail without the code added in the PR, unless the PR is specifically about improving test coverage for existing features.

File: crates/red_knot_python_semantic/src/types.rs (hunk 2 above)
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/src/types.rs:1411]: We could also just propagate the existing TODO from the left-hand side

File: crates/red_knot_python_semantic/resources/mdtest/annotations/literal_string.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+not vice versa.
+
+```py
+foo_1: Literal["foo"] = "foo"
+bar_1: LiteralString = foo_1  # fine
+
+if bool():
+    foo_2 = "foo"
+else:
+    foo_2 = "bar"
```
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/resources/mdtest/annotations/literal_string.md:112]: This is because you're using `bool()` to create the test, and red-knot knows that `bool()` with no arguments creates an object of type `Literal[False]`. It actually seems like a bug that it _doesn't_ know that for the more verbose form that doesn't use a ternary expression. We shouldn't rely on the bug.

Instead, you should create a custom function that is annotated as returning a `bool`, since red-knot will not be able to "see through" the return annotation, and use _that_ for the test:

```py
def coinflip() -> bool:
    return True

foo_2 = "foo" if coinflip() else "bar"
```

File: crates/red_knot_python_semantic/resources/mdtest/annotations/literal_string.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+qux = (foo, bar)
+reveal_type(qux)  # revealed: tuple[Literal["foo"], Literal["bar"]]
+
+# TODO: Infer "LiteralString"
+reveal_type(foo.join(qux))  # revealed: @Todo(call type)
+
+template: LiteralString = "{}, {}"
+reveal_type(template)  # revealed: Literal["{}, {}"]
+# TODO: Infer "foo, bar"
```
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply crates/red_knot_python_semantic/resources/mdtest/annotations/literal_string.md:95]: I see your point @InSyncWithFoo: maybe string templating really is special enough for us to warrant special-casing `str.format` in the way you describe here. It's a somewhat slippery slope there, though (we obviously are not building a complete Python interpreter in this project!), and I feel like it's not clear that we'll _necessarily_ add this special casing. It's also going to be a long way off; adding this kind of special casing isn't a priority for us right now. I agree with @carljm that this TODO feels unnecessary right now.

File: crates/red_knot_python_semantic/resources/mdtest/annotations/literal_string.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+TA: TypeAlias = LiteralString
+
+T1 = TypeVar("T1", bound=LiteralString)
+
+variable_annotation: LiteralString
+
+class Foo:
+    # fine: Second arguments and later must be ignored.
+    annotated: Annotated[int, Literal[LiteralString]]
```
Comment (carljm) [inline crates/red_knot_python_semantic/resources/mdtest/annotations/literal_string.md:37]:
```suggestion
from typing import LiteralString

x: LiteralString

def f():
    reveal_type(x)  # revealed: LiteralString
```

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 8]
```diff
@@ -4642,6 +4642,17 @@ impl<'db> TypeInferenceBuilder<'db> {
                 );
                 Type::Unknown
             }
+            KnownInstanceType::LiteralString => {
+                self.diagnostics.add(
+                    subscript.into(),
+                    "invalid-type-parameter",
+                    format_args!(
+                        "Type `{}` expected no type parameter. Did you mean to use `Literal[...]` instead?",
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:4650]: Nice error message!

**Review by carljm — APPROVED** [review]
Thanks!
